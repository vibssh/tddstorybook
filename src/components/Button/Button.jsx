import React from "react";
import PropTypes from "prop-types";
import "./Button.scss";

const propTypes = {
  type: PropTypes.string.isRequired,
  variance: PropTypes.string,
  size: PropTypes.string,
  state: PropTypes.string,
  btnValue: PropTypes.string
};

const defaultPropTypes = {
  variance: "default",
  size: "normal",
  state: "",
  btnValue: ""
};

const options = {
  TYPES: {
    BUTTON: "button",
    ANCHOR: "anchor",
    INPUT: "input"
  },

  VARIANCE: {
    DEFAULT: "default",
    PRIMARY: "primary",
    SECONDARY: "secondary",
    SUCCESS: "success",
    DANGER: "danger",
    WARNING: "warning",
    INFO: "info",
    LIGHT: "light",
    DARK: "dark"
  },

  SIZES: {
    LARGE: "large",
    SMALL: "small",
    NORMAL: "normal",
    BLOCK: "block"
  },

  STATES: {
    ACTIVE: "active",
    DISABLED: "disabled"
  }
};

const Button = props => {
  const { type, variance, size, state, btnValue, children } = props;
  return (
    <React.Fragment>
      {(() => {
        switch (type) {
          case options.TYPES.BUTTON:
            return (
              <button
                type="button"
                className={`btn ${variance ? variance: ''} ${state ? state: ''} ${size ? size: ''}`}
              >
                {children}
              </button>
            );
          case options.TYPES.ANCHOR:
            return (
              <a
                href="/#"
                className={`btn ${variance ? variance: ''} ${state ? state: ''} ${size ? size: ''}`}
                role="button"
              >
                {children}
              </a>
            );
          case options.TYPES.INPUT:
            return (
              <input
                type="submit"
                className={`btn ${variance ? variance: ''} ${state ? state: ''} ${size ? size: ''}`}
                value={btnValue}
              />
            );
            default: return <h1>No Button Type Declared</h1> 
        }
      })()}
    </React.Fragment>
  );
};

Button.propTypes = propTypes;
Button.defaultPropTypes = defaultPropTypes;

export default Button;
