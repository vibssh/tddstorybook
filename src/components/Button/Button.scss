/* Zero off Button native browser styles */
button {
  -webkit-appearance: none;
  border: none;
}

/* Colors for Button component */
$default: #444444;
$primary: #007bff;
$secondary: #6c757d;
$success: #28a745;
$danger: #dc3545;
$warning: #ffc107;
$info: #17a2b8;
$light: #f8f9fa;
$dark: #343a40; 

.btn {
  display: inline-block;
  padding: 0.375rem .75rem;
  vertical-align: middle;
  text-align: center;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  font-weight: 400;
  color: #ffffff;
  background: $default;
  font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
  text-decoration: none;
  border: none;

  &.primary {
    background: $primary;
  }

  &.secondary {
    background: $secondary;
  }

  &.success {
    background: $success;
  }

  &.danger {
    background: $danger;
  }

  &.warning {
    background: $warning;
  }

  &.info {
    background: $info;
  }

  &.light {
    background: $light;
    color: $dark;
  }

  &.dark {
    background: $dark;
    color: $light;
  }
}