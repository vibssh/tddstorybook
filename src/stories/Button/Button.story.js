import React from "react";
import { storiesOf } from "@storybook/react";
import Button from "../../components/Button";

const Types = ["button", "anchor", "input"];
const Variance = [
  "default",
  "primary",
  "secondary",
  "danger",
  "success",
  "warning",
  "info",
  "light",
  "dark"
];

const iterator = () => {
  Types.map((type) => (
    Variance.map((v) => {
      const storyTitle = (type === 'button') ? '' : (type === 'button' && v === 'default') ? 'Default' : type.charAt(0).toUpperCase() + type.slice(1);
      return (
        storiesOf("Button", module)
        .addWithJSX(`${storyTitle} ${v !== 'default' ? v.charAt(0).toUpperCase() + v.slice(1) : ''} Button`, () => {
        return <Button type={type} variance={v}>{`${v.charAt(0).toUpperCase() + v.slice(1)} Button`}</Button>;
      })
    )
    })
  ))
}
iterator();
