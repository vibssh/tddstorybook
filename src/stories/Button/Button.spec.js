import React from "react";
import { shallow } from "enzyme";
import Button from "../../components/Button";
const options = {
  TYPES: {
    BUTTON: "button",
    ANCHOR: "anchor",
    INPUT: "input"
  },

  VARIANCE: {
    DEFAULT: "default",
    PRIMARY: "primary",
    SECONDARY: "secondary",
    SUCCESS: "success",
    DANGER: "danger",
    WARNING: "warning",
    INFO: "info",
    LIGHT: "light",
    DARK: "dark"
  },

  SIZES: {
    LARGE: "large",
    SMALL: "small",
    NORMAL: "normal",
    BLOCK: "block"
  },

  STATES: {
    ACTIVE: "active",
    DISABLED: "disabled"
  }
};

describe("Button", () => {
  let wrapper;
  it("should render a default button", () => {
    wrapper = shallow(
      <Button type={options.TYPES["BUTTON"]}>Default Button</Button>
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });
  describe("Primary Button", () => {
    it("should render a Primary button", () => {
      wrapper = shallow(
        <Button
          type={options.TYPES["BUTTON"]}
          variance={options.VARIANCE["PRIMARY"]}
        >
          {" "}
          Primary Button{" "}
        </Button>
      );
      expect(wrapper.debug()).toMatchSnapshot();
    });
    
  });
});
