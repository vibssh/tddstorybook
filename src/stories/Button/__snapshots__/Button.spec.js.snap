// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Primary Button should render a Primary button 1`] = `
"<Fragment>
  <button type=\\"button\\" className=\\"btn primary  \\">
     
    Primary Button
     
  </button>
</Fragment>"
`;

exports[`Button should render a default button 1`] = `
"<Fragment>
  <button type=\\"button\\" className=\\"btn   \\">
    Default Button
  </button>
</Fragment>"
`;
